src/data/finset/sigma.lean

```diff
@@ -55,6 +66,70 @@ end
 
 lemma inf_sigma [semilattice_inf β] [order_top β] :
   (s.sigma t).inf f = s.inf (λ i, (t i).inf $ λ b, f ⟨i, b⟩) :=
-@sup_sigma _ (order_dual β) _ _ _ _ _ _
+@sup_sigma _ _ (order_dual β) _ _ _ _ _
+
+end sigma
+
+section sigma_lift
+variables {α β γ : ι → Type*} [decidable_eq ι]
+
+/-- Lifts maps `α i → β i → finset (γ i)` to a map `Σ i, α i → Σ i, β i → finset (Σ i, γ i)`. -/
+def sigma_lift (f : Π ⦃i⦄, α i → β i → finset (γ i)) (a : sigma α) (b : sigma β) :
+  finset (sigma γ) :=
+dite (a.1 = b.1) (λ h, (f (h.rec a.2) b.2).map $ embedding.sigma_mk _) (λ _, ∅)
+
+lemma mem_sigma_lift (f : Π ⦃i⦄, α i → β i → finset (γ i))
+  (a : sigma α) (b : sigma β) (x : sigma γ) :
+  x ∈ sigma_lift f a b ↔ ∃ (ha : a.1 = x.1) (hb : b.1 = x.1), x.2 ∈ f (ha.rec a.2) (hb.rec b.2) :=
```

Comment:
The right hand side of the iff is quite terrifying, so I'd add unbundled versions as follows:
```
lemma mem_sigma_lift_eq (f : Π ⦃i⦄, α i → β i → finset (γ i))
  (i : ι) (a : α i) (b : β i) (x : γ i) :
  (i, x) ∈ sigma_lift f (i, a) (i, b) ↔ x ∈ f a b := sorry

lemma mem_sigma_lift_ne_left (f : Π ⦃i⦄, α i → β i → finset (γ i))
  (i₁ i₂ j : ι) (hij : i₁ ≠ j) (a : α i₁) (b : β i₂) (x : γ j) :
  (j, x) \notin sigma_lift f (i, a) (i, b) := sorry

lemma mem_sigma_lift_ne_right (f : Π ⦃i⦄, α i → β i → finset (γ i))
  (i₁ i₂ j : ι) (hij : i₂ ≠ j) (a : α i₁) (b : β i₂) (x : γ j) :
  (j, x) \notin sigma_lift f (i, a) (i, b) := sorry
```

Reply:
The first one is nice, but I don't think the last two are worth the unbundling. What do you think of the two I put instead?